src/zip.c

```diff
@@ -1023,6 +1081,19 @@ int zip_extract_stream(const char *stream, size_t size, const char *dir,
   return status;
 }
 
+typedef enum {
+  KEEP = 0,
+  DELETE = 1,
+  MOVE = 2,
+} modify_type;
+
+struct entry_mark {
+  int file_index;
+  modify_type type;
+  mz_uint64 m_local_header_ofs;
+  mz_uint64 lf_length;
+};
+
```

Comment:
When I want to  delete one folder (multiple files) at a time,  I need to  record the information about all files to be deleted and then delete them one by one. The struct entry_mark is for this perpose. An entry_mark element keeps the information of a file, An array of entry_mark element keeps the  information of all files in .ZIP file. The index of entry_mark element in the array shows the occurrence order of a file in  central directory structure, the variables in the structure are described as follows:

`struct entry_mark {`
  `   int file_index;     // shows the order in which local file header related with this entry  appears  `
  `   modify_type type;    //  shows the file related with this entry should be kept, deleted, or moved?`
   `  mz_uint64 m_local_header_ofs;  // keeps the offset of file data(local file header + file data + data descriptor )  relative to the start of the .ZIP file`
  `   mz_uint64 lf_length;  //  keeps the file data(local file header + file data + data descriptor ) length`
`};`

Reply:
By the way, this pr supports the deletion of multiple files at a time a at the bottom level, but it has not been implemented at the interface level, I'm working on it.

Reply:
Maybe we can simplify it and expose api which let you delete entries one by one.
I'm not 100% sure, but if you take a look into zip command (delete entries parameter), then for zip file:
```
zipinfo tmp.zip 
Archive:  tmp.zip
Zip file size: 1378 bytes, number of entries: 10
drwxr-xr-x  3.0 unx        0 bx stor 20-Oct-25 02:04 a/
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/a3
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/a2
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/a1
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:04 a/c2
drwxr-xr-x  3.0 unx        0 bx stor 20-Oct-25 02:05 b/
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:05 b/b2
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:05 b/b3
-rw-r--r--  3.0 unx        0 bx stor 20-Oct-25 02:05 b/b1
drwxr-xr-x  3.0 unx        0 bx stor 20-Oct-25 02:04 empty/

zip tmp.zip -d "a/"
deleting: a/

zip tmp.zip -d "a/*"
deleting: a/a3
deleting: a/a2
deleting: a/a1
deleting: a/c2
```

I'm not saying to implement _regexp_ like `dir/*`, but maybe for simplicity `dir/` will remove just this particular entry name.
WDYT?

Reply:
The current pr supports to delete a file or a folder at a time. The process of the current PR is as follows:

1. Open .zip file, initialize a data structure `mz_zip_archive`. 
2. create an array `entry_mark_array` to keep the information(data offset, data length, if the data needs to be deleted, the mapping relationship between the entries in the central directory and the file data blocks) about all compressed files 
3. Traverse `entry_mark_array`, delete all file data and modify the data which is affected by the deleted file in the central directory in one traversal. Then delete the entrys in the central directory and truncate the file at the end.
(If the .zip file contains a dir/ and dir/* , we can't delete the entry dir/ alone now, this is a little bug, I can fix it and implement regexp like dir/* by modify several lines of code. )

If we just creat a simple api which let people delete entries one by one. I think the api process can be as follows(there may be a simpler way to do it):

1. Open .zip file, initialize a data structure `mz_zip_archive`.
2. Get the information that needed for deleting a file. The information includes the length and the offset relative to the file start position of the data block to be deleated . then modify the data which is affected by the deleted file in the central directory.
 3. move the data block behind the deleated data block forward to overlay the deleated data block,  and truncate the file at the end.

   In step 2, we can  call the function `zip_entry_open(zip, deleted_file_name)` to get some information that needed for deleting a file. We can only get `entry index`, `m_central_dir_ofs` and `m_local_header_ofs`. to get data length we have other work to do. Since the  [.ZIP file format spec](https://pkware.cachefly.net/webdocs/APPNOTE/APPNOTE-6.2.0.txt) says `"The entries in the central directory may not necessarily be in the same order that files appear in the .ZIP file"`, to delete data correctly from a .zip file, we still need to create a data structure to keep the index, offset and length of all file data blocks(local file header + file data + data descriptor).what's more, we need to record the association between the entries in the central directory and the file data blocks. 

  So in my opinion, we have limited space for simplifying the code if we just expose a simple api. On the other hand, we will waste a lot of time in initializing auxiliary data structures when deleting folders by using a simple API, especially when there are many files in the folder. I think it would be very useful to support deleting folders. But if we have a special need to implment a simple api, I can do it too.

Reply:
@kuba-- 
I've updated the code and  improved the current API. Although I know you want a simple API, not a complex API with full functions. But it's hard to reduce a lot of code when simpling the current API.  So I just made a little change to implement the regexp.  Do you have any ideas for implementing a simple API?